### PR TITLE
gh-112075: Adapt more dict methods to Argument Clinic

### PR DESCRIPTION
**Pull request by DinoV**
Uses argument clinic for the remaining `dict` methods which aren't using it yet.  This will allow to use argument clinic's `@critical_section` consistently where it's appropriate.

Mostly straight forward, the most dramatic thing is that some of the argument clinic names overlap with existing names.  Those existing names are just wrapped in public dict API so the functionality just moves into the public API.

* Issue: gh-112075

### CODE REVIEW

File: Objects/dictobject.c [hunk 1]
```diff
@@ -3108,10 +3123,17 @@ _PyDict_MergeEx(PyObject *a, PyObject *b, int override)
     return dict_merge(interp, a, b, override);
 }
 
+/*[clinic input]
+dict.copy
+
+D.copy() -> a shallow copy of D
```
Comment (methane) [inline Objects/dictobject.c]:
```suggestion
Returns a shallow copy of the dictionary.
```

File: Objects/dictobject.c [hunk 2]
```diff
@@ -3512,10 +3504,18 @@ dict_setdefault_impl(PyDictObject *self, PyObject *key,
     return Py_XNewRef(val);
 }
 
+
+/*[clinic input]
+dict.clear
+
+D.clear() -> None.  Remove all items from D.
```
Comment (methane) [inline Objects/dictobject.c]:
```suggestion
Removes all items.
```

File: Objects/dictobject.c [hunk 3]
```diff
@@ -3703,11 +3703,17 @@ _PyDict_KeysSize(PyDictKeysObject *keys)
     return size;
 }
 
+/*[clinic input]
+dict.__sizeof__
+
+D.__sizeof__() -> size of D in memory, in bytes
```
Comment (methane) [inline Objects/dictobject.c]:
```suggestion
Returns size in memory, in bytes.
```

File: Objects/dictobject.c [hunk 4]
```diff
@@ -5368,10 +5356,17 @@ PyTypeObject PyDictItems_Type = {
     .tp_getset = dictview_getset,
 };
 
+/*[clinic input]
+dict.items
+
+D.items() -> a set-like object providing a view on D's items
```
Comment (methane) [inline Objects/dictobject.c]:
```suggestion
Returns a set-like object providing a view on items.
```

File: Objects/dictobject.c [hunk 5]
```diff
@@ -5451,10 +5446,17 @@ PyTypeObject PyDictValues_Type = {
     .tp_getset = dictview_getset,
 };
 
+/*[clinic input]
+dict.values
+
+D.values() -> an object providing a view on D's values
```
Comment (methane) [inline Objects/dictobject.c]:
```suggestion
Returns an object providing a view on values.
```

File: Objects/dictobject.c [hunk 6]
```diff
@@ -5263,10 +5244,17 @@ PyTypeObject PyDictKeys_Type = {
     .tp_getset = dictview_getset,
 };
 
+/*[clinic input]
+dict.keys
+
+D.keys() -> a set-like object providing a view on D's keys
```
Comment (methane) [inline Objects/dictobject.c]:
```suggestion
Returns a set-like object providing a view on keys.
```

File: Objects/dictobject.c [hunk 7]
```diff
@@ -3108,10 +3123,17 @@ _PyDict_MergeEx(PyObject *a, PyObject *b, int override)
     return dict_merge(interp, a, b, override);
 }
 
+/*[clinic input]
+dict.copy
+
+Returns a shallow copy of D.
```
Comment (methane) [inline Objects/dictobject.c]: What is `D`?
"Returns a shallow copy." or "Returns a shallow copy of the dict." would be better.

File: Objects/dictobject.c [hunk 8]
```diff
@@ -3703,11 +3703,17 @@ _PyDict_KeysSize(PyDictKeysObject *keys)
     return size;
 }
 
+/*[clinic input]
+dict.__sizeof__
+
+Returns size of D in memory, in bytes.
```
Comment (methane) [inline Objects/dictobject.c]: ditto. Don't use `D` when converting docstring to clinic.

File: Objects/dictobject.c [hunk 9]
```diff
@@ -5368,10 +5356,17 @@ PyTypeObject PyDictItems_Type = {
     .tp_getset = dictview_getset,
 };
 
+/*[clinic input]
+dict.items
+
+Returns a set-like object providing a view on D's items.
```
Comment (methane) [inline Objects/dictobject.c]: ditto

File: Objects/dictobject.c [hunk 10]
```diff
@@ -5451,10 +5446,17 @@ PyTypeObject PyDictValues_Type = {
     .tp_getset = dictview_getset,
 };
 
+/*[clinic input]
+dict.values
+
+Returns an object providing a view on D's values.
```
Comment (methane) [inline Objects/dictobject.c]: ditto

File: Objects/dictobject.c [hunk 11]
```diff
@@ -3739,56 +3745,31 @@ dict_ior(PyObject *self, PyObject *other)
...
-PyDoc_STRVAR(values__doc__,
-             "D.values() -> an object providing a view on D's values");
 
 static PyMethodDef mapp_methods[] = {
     DICT___CONTAINS___METHODDEF
     {"__getitem__",     dict_subscript,                 METH_O | METH_COEXIST,
      getitem__doc__},
-    {"__sizeof__",      dict_sizeof,                    METH_NOARGS,
-     sizeof__doc__},
+     DICT___SIZEOF___METHODDEF
```
Comment (erlend-aasland) [inline Objects/dictobject.c]:
```suggestion
    DICT___SIZEOF___METHODDEF
```

File: Objects/dictobject.c [hunk 12]
```diff
@@ -3703,11 +3703,17 @@ _PyDict_KeysSize(PyDictKeysObject *keys)
     return size;
 }
 
+/*[clinic input]
+dict.__sizeof__
+
+Returns size of the dict in memory, in bytes.
```
Comment (erlend-aasland) [inline Objects/dictobject.c]:
```suggestion
Return the size of the dict in memory, in bytes.
```

File: Objects/dictobject.c [hunk 13]
```diff
@@ -3512,10 +3504,18 @@ dict_setdefault_impl(PyDictObject *self, PyObject *key,
     return Py_XNewRef(val);
 }
 
+
+/*[clinic input]
+dict.clear
+
+Removes all items.
```
Comment (erlend-aasland) [inline Objects/dictobject.c]:
```suggestion
Remove all items from the dict.
```

File: Objects/dictobject.c [hunk 14]
```diff
@@ -3108,10 +3123,17 @@ _PyDict_MergeEx(PyObject *a, PyObject *b, int override)
     return dict_merge(interp, a, b, override);
 }
 
+/*[clinic input]
+dict.copy
+
+Returns a shallow copy of the dict.
```
Comment (erlend-aasland) [inline Objects/dictobject.c]:
```suggestion
Return a shallow copy of the dict.
```

File: Objects/dictobject.c [hunk 15]
```diff
@@ -5263,10 +5244,17 @@ PyTypeObject PyDictKeys_Type = {
     .tp_getset = dictview_getset,
 };
 
+/*[clinic input]
+dict.keys
+
+Returns a set-like object providing a view on the dict's keys.
```
Comment (erlend-aasland) [inline Objects/dictobject.c]:
```suggestion
Return a set-like object providing a view on the dict's keys.
```

File: Objects/dictobject.c (hunk 9 above)
Comment (erlend-aasland) [inline Objects/dictobject.c]:
```suggestion
Return a set-like object providing a view on the dict's items.
```

File: Objects/dictobject.c [hunk 16]
```diff
@@ -5451,10 +5446,17 @@ PyTypeObject PyDictValues_Type = {
     .tp_getset = dictview_getset,
 };
 
+/*[clinic input]
+dict.values
+
+Returns an object providing a view on the dict's values.
```
Comment (erlend-aasland) [inline Objects/dictobject.c]:
```suggestion
Return an object providing a view on the dict's values.
```

File: Objects/dictobject.c (hunk 1 above)
Comment (erlend-aasland) [reply Objects/dictobject.c]: We use imperative mood for the reference; I added suggestions to amend the docstrings.

**Comment by erlend-aasland** [comment]
Sorry to add remarks on the already amended docstrings, but let's avoid the CI churn of addressing this in another PR :)

**Review by erlend-aasland — APPROVED** [review]
LGTM with last round of suggestions applied.

**Comment by DinoV** [comment]
Thanks, I've appied them all!

**Comment by erlend-aasland** [comment]
> Thanks, I've appied them all!

Regen clinic ;)

**Comment by erlend-aasland** [comment]
@DinoV, next time, can you please clean up the commit message before merging?

**Comment by DinoV** [comment]
> @DinoV, next time, can you please clean up the commit message before merging?

Ahh yes, sorry! :(

**Pull request merged**